### PR TITLE
ARROW-13999: [C++] Fix bundled LZ4 build on MinGW

### PR DESCRIPTION
**Pull request by lidavidm**
This doesn't fix the installed LZ4 not being detected in the first place, however.

### CODE REVIEW

**Comment by lidavidm** [comment]
By comparing with the libzstd finder (which succeeds) the difference appears to be the following: 

In libzstd.pc, we have `libdir=${prefix}/lib` while in liblz4.pc, we have `libdir=/mingw64/lib`. It appears CMake can only properly search the former path, which resolves to a Windows-style path, not one of MinGW's fake Unix paths.

**Comment by lidavidm** [comment]
Comparing the packages, pkg-config did change. The commit is https://github.com/msys2/MINGW-packages/commit/4861bb970cb99879939033fe9cfbcfc841d49402 but it seems this shouldn't have an effect? The patch removed there was upstreamed in 1.8.0.

**Comment by lidavidm** [comment]
Building and installing pkg-config 1.7 in MinGW also fixes it, so there's likely some regression in 1.8?

File: cpp/build-support/build-lz4-lib.sh [hunk 1]
```diff
@@ -17,9 +17,14 @@
 # specific language governing permissions and limitations
 # under the License.
 #
+MAKE="make"
+if [ -n "$MSYSTEM" ]; then
+  MAKE="mingw32-make"
+fi
```
Comment (kou) [inline cpp/build-support/build-lz4-lib.sh:23]: How about setting correct `PATH` such as `PATH=/mingw32/bin:$PATH` by user instead of this?

Comment (lidavidm) [reply cpp/build-support/build-lz4-lib.sh:23]: I found that the mingw-w64-x86_64-make package doesn't install `make` but rather `mingw32-make`, unfortunately. (And from what I understand, the msys2 make package shouldn't be mixed with the mingw toolchain we use here.)

**Comment by lidavidm** [comment]
I filed https://github.com/pkgconf/pkgconf/issues/234 that said, it may not be a pkgconf issue per se but rather an issue with the .pc files that the pkgconf upgrade exposed.

File: cpp/build-support/build-lz4-lib.sh (hunk 1 above)
Comment (kou) [reply cpp/build-support/build-lz4-lib.sh:23]: Oh, I see.
I confirmed it at: https://packages.msys2.org/package/mingw-w64-x86_64-make?repo=mingw64

**Comment by lidavidm** [comment]
@kou how does the fix here look? I added some clarifications. Hopefully this can get fixed upstream in msys/pkgconf so we can avoid building LZ4 ourselves, but for now this will allow CI to pass.

**Comment by kou** [comment]
Recent LZ4 supports CMake build: https://github.com/lz4/lz4/tree/dev/build/cmake
How about using it instead of raw `make` build support? Then we don't need to add this workaround.

**Comment by lidavidm** [comment]
Ah! That would be much better. I'll take a look tomorrow.

**Comment by lidavidm** [comment]
Actually, does that make sense? The CMake build support isn't in any LZ4 release; we're already on the latest, v1.9.3, which doesn't include that yet. Do we want to depend on an unreleased LZ4 version?

**Comment by pitrou** [comment]
The CMake support in LZ4 is probably quite young and my fear is that it will take some time to stabilize. I think this PR is fine in the meantime, and we can open a followup JIRA to switch to CMake in e.g. 6 months.

**Comment by nealrichardson** [comment]
Jira made: ARROW-14015

I'll merge this now to unblock some other PRs.

**Comment by kou** [comment]
FYI: CMake support is included in 1.9.3:

```console
$ find lz4-1.9.3 | grep CMakeLists.txt
lz4-1.9.3/build/cmake/CMakeLists.txt
```

https://github.com/lz4/lz4/archive/v1.9.3.tar.gz

**Comment by lidavidm** [comment]
Ah, I'm sorry, I only looked at the latest commit for some reason >.< I'll try to pick up that ticket at some point then.

(Note that msys2 upstream also patched their pkgconfig so the underlying issue should be resolved for now.)

**Pull request closed without merging**